test/tools/router_check/router.cc

```diff
@@ -336,59 +391,118 @@ bool RouterCheckTool::compareRedirectPath(
   return compareRedirectPath(tool_config, expected.path_redirect().value());
 }
 
-bool RouterCheckTool::compareRequestHeaderField(
+bool RouterCheckTool::compareRequestHeaderFields(
     ToolConfig& tool_config, const envoy::RouterCheckToolSchema::ValidationAssert& expected) {
   bool no_failures = true;
+  if (expected.request_header_matches().data()) {
+    for (const envoy::config::route::v3::HeaderMatcher& header :
+         expected.request_header_matches()) {
+      if (!matchHeaderField(*tool_config.request_headers_, header, "request_header_matches")) {
+        no_failures = false;
+      }
+    }
+  }
+  // TODO(kb000) : Remove deprecated request_header_fields.
   if (expected.request_header_fields().data()) {
     for (const envoy::config::core::v3::HeaderValue& header : expected.request_header_fields()) {
-      if (!compareRequestHeaderField(tool_config, header.key(), header.value())) {
+      if (!compareHeaderField(*tool_config.request_headers_, header.key(), header.value(),
+                              "request_header_fields", true)) {
         no_failures = false;
       }
     }
   }
   return no_failures;
 }
 
-bool RouterCheckTool::compareRequestHeaderField(ToolConfig& tool_config, const std::string& field,
-                                                const std::string& expected) {
-  std::string actual = tool_config.request_headers_->get_(field);
-  return compareResults(actual, expected, "request_header_fields");
-}
-
-bool RouterCheckTool::compareResponseHeaderField(
+bool RouterCheckTool::compareResponseHeaderFields(
     ToolConfig& tool_config, const envoy::RouterCheckToolSchema::ValidationAssert& expected) {
   bool no_failures = true;
+  if (expected.response_header_matches().data()) {
+    for (const envoy::config::route::v3::HeaderMatcher& header :
+         expected.response_header_matches()) {
+      if (!matchHeaderField(*tool_config.response_headers_, header, "response_header_matches")) {
+        no_failures = false;
+      }
+    }
+  }
+  // TODO(kb000) : Remove deprecated response_header_fields.
   if (expected.response_header_fields().data()) {
     for (const envoy::config::core::v3::HeaderValue& header : expected.response_header_fields()) {
-      if (!compareResponseHeaderField(tool_config, header.key(), header.value())) {
+      if (!compareHeaderField(*tool_config.response_headers_, header.key(), header.value(),
+                              "response_header_fields", /*expect_match=*/true)) {
         no_failures = false;
       }
     }
   }
   return no_failures;
 }
 
-bool RouterCheckTool::compareResponseHeaderField(ToolConfig& tool_config, const std::string& field,
-                                                 const std::string& expected) {
-  std::string actual = tool_config.response_headers_->get_(field);
-  return compareResults(actual, expected, "response_header_fields");
+template <typename HeaderMapType>
+bool RouterCheckTool::matchHeaderField(const HeaderMapType& header_map,
```

Comment:
Can't you just use `HeaderUtility::matchHeaders()` etc. to do this instead of reimplementing all of this? I'm unclear why we have to have this here again. Then you support all of the fields for free and can delete all of the special casing in the docs.

Reply:
The implementation here writes logs that are useful for route validation. `HeaderUtility` lacks this.

Reply:
I think it would be better to use the real implementation for the validation, and then add logging on top if you need that somehow.

/wait